### PR TITLE
 BUG: iloc setting columns not taking effect

### LINKED ISSUE #33198: BUG: iloc setting columns not taking effect
```
arr = np.random.randn(10 ** 6).reshape(500, 2000).astype(np.float64)
df = pd.DataFrame(arr)
df.iloc[:, 1000:] = df.iloc[:, 1000:].astype(np.float32)

>>> df.dtypes.value_counts()
float64    2000
dtype: int64
```

I would expect 1000 float64 columns and 1000 float32 columns.  I get the expected behavior if I set a single column before trying to set all 1000:

```
df = pd.DataFrame(arr)
df[1000] = df[1000].astype(np.float32)
df.iloc[:, 1000:] = df.iloc[:, 1000:].astype(np.float32)

>>> df.dtypes.value_counts()
float32    1000
float64    1000
dtype: int64
```

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [x] closes #33198
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
are you running the full test suite locally before pushing?

**Comment by MomIsBestFriend** [comment]
> are you running the full test suite locally before pushing?

Not fully. usually only the parts I suspect that are related to the changes I make, in this case I tried on ```pytest -q --cache-clear pandas/tests/frame/methods/``` before pushing.

**Comment by jbrockmendel** [comment]
Unless you need to save the wattage, its usually good to run the full thing.  e.g. i have in .bash_aliases `alias ptest="python3 -m pytest pandas/tests --skip-slow --skip-db"`

In this case we might expect tests.indexing, tests.frame.indexing, tests.series.indexing, and possibly tests.indexes to be potential trouble spots

**Comment by MomIsBestFriend** [comment]
> Unless you need to save the wattage, its usually good to run the full thing. e.g. i have in .bash_aliases `alias ptest="python3 -m pytest pandas/tests --skip-slow --skip-db"`

That's a good tip.

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1557,7 +1557,10 @@ def _setitem_with_indexer(self, indexer, value):
             (blk,) = self.obj._mgr.blocks
             if 1 < blk.ndim:  # in case of dict, keys are indices
                 val = list(value.values()) if isinstance(value, dict) else value
-                take_split_path = not blk._can_hold_element(val)
+                if isinstance(indexer, tuple) and com.is_null_slice(indexer[0]):
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:1560]: can you add a comment pointing back to this or the original issue

File: pandas/tests/frame/indexing/test_indexing.py [hunk 2]
```diff
@@ -31,6 +31,15 @@
 _slice_msg = "slice indices must be integers or None or have an __index__ method"
 
 
+def test_unsure_about_the_name_and_location():
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_indexing.py:34]: we need to make a test_iloc.py file in this directory, might as well do it in this PR

**Comment by jbrockmendel** [comment]
Looking at the test logs, there's still some more troubleshooting to be done.  What's the next step?

**Comment by MomIsBestFriend** [comment]
> Looking at the test logs, there's still some more troubleshooting to be done. What's the next step?

@jbrockmendel Looks like we need to check for the length of the index of each one, to decide if we are taking the split path.

what I tried was:

```python
                val = list(value.values()) if isinstance(value, dict) else value
                # https://github.com/pandas-dev/pandas/issues/33198
                if isinstance(indexer, tuple) and com.is_null_slice(indexer[0]):
                    take_spit_path = length_of_indexer(
                        self.obj.index
                    ) == length_of_indexer(val.index)
```

---

But it does not work.

**Comment by jbrockmendel** [comment]
Let's continue troubleshooting in #33198.  Closing to clear the queue, can re-open when its ready to go.

**Pull request closed without merging**